fastcrypto-tbls/src/dkg.rs

```diff
@@ -0,0 +1,404 @@
+                let buff = bincode::serialize(&share.value)
+                    .expect("serialize of a share should never fail");
+                let encryption = n.pk.encrypt(&buff, rng);
+                EncryptedShare {
+                    receiver: n.id,
+                    encryption,
+                }
+            })
+            .collect();
+
+        FirstMessage {
+            sender: self.id,
+            encrypted_shares,
+            vss_pk: self.vss_pk.clone(),
+        }
+    }
+
+    /// 5. Process the first messages of exactly 'threshold' nodes and create the second message to
+    ///    be broadcast.
+    ///    The second message contains the list of complaints on invalid shares. In addition, it
+    ///    returns a set of valid shares (so far).
+    ///    Since we assume that at most t-1 of the nodes are malicious, we only need messages from
+    ///    t nodes to guarantee an unbiasable and unpredictable beacon. (The result is secure with
+    ///    rushing adversaries according to https://eprint.iacr.org/2021/005.pdf.)
+    pub fn create_second_message<R: AllowedRng>(
+        &self,
+        messages: &[FirstMessage<G, EG>],
+        rng: &mut R,
+    ) -> Result<(SharesMap<G>, SecondMessage<EG>), FastCryptoError> {
+        if messages.len() != self.threshold as usize {
+            return Err(FastCryptoError::InputLengthWrong(self.threshold as usize));
+        }
+
+        let my_id = self.id;
+        let mut shares = HashMap::new(); // Will include only valid shares.
+        let mut next_message = SecondMessage {
+            sender: my_id,
+            complaints: Vec::new(),
+        };
+
+        for message in messages {
+            // Ignore if the threshold is different (and other honest parties will ignore as well).
+            if message.vss_pk.degree() != self.threshold - 1 {
+                continue;
+            }
+            if message.sender == my_id {
+                shares.insert(message.sender, self.vss_sk.eval(my_id).value);
+                continue;
+            }
+            // TODO: check that current dealer is in the list of pki nodes.
+            // Get the relevant encrypted share (or skip message).
+            let encrypted_share = message
+                .encrypted_shares
+                .iter()
+                .find(|n| n.receiver == my_id);
+            // No share for me.
+            if encrypted_share.is_none() {
+                next_message
+                    .complaints
+                    .push(Complaint::NoShare(message.sender));
+                continue;
+            }
+            // Else, decrypt it.
+            let share = Self::decrypt_and_check_share(
+                &self.ecies_sk,
+                my_id,
+                &message.vss_pk,
+                encrypted_share.expect("checked above that is not None"),
+            );
+            match share {
+                Ok(sh) => {
+                    shares.insert(message.sender, sh);
```

Comment:
do we need to detect messages from the same sender to avoid duplicates or we expect that the last insert always wins? Generally, as long as we use vectors vs hashmaps for to receive messages as function input, how do we deal with duplicated data at this stage (step)? in general do we perform a fail fast approach or just ignore duplicates?

Reply:
added a check